### ISSUE TITLE
Too Many Connects problems

### ISSUE THREAD

**Issue by chubbard** [aws/aws-msk-iam-auth#45]
I'm using a small sized broker (2), and I'm experiencing the **too many connects** problem:

    org.apache.kafka.common.errors.SaslAuthenticationException: [c4c1210b-9221-40ac-9fde-03daf7b747dc]: Too many connects

    java.lang.IllegalStateException: This error handler cannot process 'org.apache.kafka.common.errors.SaslAuthenticationException's; no record information is available
	at org.springframework.kafka.listener.SeekUtils.seekOrRecover(SeekUtils.java:200)
	at org.springframework.kafka.listener.SeekToCurrentErrorHandler.handle(SeekToCurrentErrorHandler.java:112)
	at org.springframework.kafka.listener.KafkaMessageListenerContainer$ListenerConsumer.handleConsumerException(KafkaMessageListenerContainer.java:1602)
	at org.springframework.kafka.listener.KafkaMessageListenerContainer$ListenerConsumer.run(KafkaMessageListenerContainer.java:1210)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run$$$capture(FutureTask.java:264)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java)
	at java.base/java.lang.Thread.run(Thread.java:834)
    Caused by: org.apache.kafka.common.errors.SaslAuthenticationException: [c4c1210b-9221-40ac-9fde-03daf7b747dc]: Too many connects

I've also adjusted my properties to raise up the reconnect rate to 1000 as suggested in the FAQ, but I'm still experiencing the same problem.

Here are my settings:

    # MSK AWS setup
    spring.kafka.security.protocol=SASL_SSL
    spring.kafka.properties.sasl.mechanism=AWS_MSK_IAM
    spring.kafka.properties.sasl.jaas.config = software.amazon.msk.auth.iam.IAMLoginModule required;
    spring.kafka.properties.sasl.client.callback.handler.class = software.amazon.msk.auth.iam.IAMClientCallbackHandler

    #spring.kafka.producer.properties.reconnect.backoff.max.ms=5000
    spring.kafka.producer.properties.reconnect.backoff.ms=1000

    #spring.kafka.consumer.properties.reconnect.backoff.max.ms=5000
    spring.kafka.consumer.properties.reconnect.backoff.ms=1000

    #spring.kafka.streams.properties.reconnect.backoff.max.ms=5000
    spring.kafka.streams.properties.reconnect.backoff.ms=1000

It doesn't change anything if I comment/uncomment the backoff.max properties.  I still get **too many connects**.

I'm seeing some strange output from MSK agent (but maybe this is ok because Kafka doesn't recognize it, but **aws-msk-iam-auth** will?

    {
     "timestamp" : "2021-10-13 10:17:33.053",
     "level" : "WARN",
     "thread" : "main",
     "logger" : "org.apache.kafka.clients.admin.AdminClientConfig",
     "message" : "The configuration 'ssl.truststore.location' was supplied but isn't a known config.",
     "context" : "default"
    }
    {
      "timestamp" : "2021-10-13 10:17:33.056",
      "level" : "WARN",
      "thread" : "main",
      "logger" : "org.apache.kafka.clients.admin.AdminClientConfig",
      "message" : "The configuration 'sasl.jaas.config' was supplied but isn't a known config.",
      "context" : "default"
    }
    {
      "timestamp" : "2021-10-13 10:17:33.056",
      "level" : "WARN",
      "thread" : "main",
      "logger" : "org.apache.kafka.clients.admin.AdminClientConfig",
      "message" : "The configuration 'sasl.client.callback.handler.class' was supplied but isn't a known config.",
      "context" : "default"
    }
    {
      "timestamp" : "2021-10-13 10:17:33.203",
      "level" : "DEBUG",
      "thread" : "main",
      "logger" : "software.amazon.msk.auth.iam.internals.MSKCredentialProvider",
      "message" : "Number of options to configure credential provider 0",
      "context" : "default"
    }

So how do I get past this ***too many connects*** issue?

**Comment by chubbard** [comment]
So I cranked it up to 3000, but same issue.  I am using a t3.small as the broker type.  So do the small brokers support 20, and the M5.large supports > 20?  Or are you saying M5.large support 20, and t3.small < 20?  I was under the impression small would support 20 connects / s.

So I only have 1 spring boot process connecting as a consumer, but that process probably has several threads as I see several thread names in the log statements.  I also have 4 topics with 2  partitions and 3 replicas that this consumer is listening and publishing to.   Some of the topics are listened to by registering @KafkaListener, and some of those are using StreamBuilder to listen to.  So that means this process carries consumers, producers, and streams clients.  I'm using default concurrency settings, except for 1 topic which I set to concurrency of 1 which if I had to guess #1-0-C-1 might be the thread listening to that topic.

I found the following threads.  The ones with asterisks (*) next to them experienced a "too many connections" exception:

main *
org.springframework.kafka.KafkaListenerEndpointContainer#0-0-C-1 *
org.springframework.kafka.KafkaListenerEndpointContainer#0-1-C-1 *
org.springframework.kafka.KafkaListenerEndpointContainer#1-0-C-1 *
org.springframework.kafka.KafkaListenerEndpointContainer#2-0-C-1 *
org.springframework.kafka.KafkaListenerEndpointContainer#2-1-C-1 *
agent-StreamThread-1 *
kafka-producer-network-thread | agent-StreamThread-1-producer
kafka-producer-network-thread | agent-producer-1

I'm left to infer that the threads that experienced the exceptions are directly communicating with the brokers, and are not simply background queue consumers used for processing callbacks.  So that leaves at least 7 threads trying to open connections to the brokers.  I'm sure the other 2 threads probably do communicate, but just got lucky and never received an exception, or never received any messages due to the other exceptions.  But under the covers of Spring Kafka integration I'm not sure how threads map to producers, etc.  Seems like sharing an open connection would make more sense than a thread / topic / worker.

In my situation I have a single big job that comes in initially, and the agent breaks down that job into millions of messages to process along some grouping rules to limit concurrency in processing the data in the correct order.  Right now I'm testing with much smaller data sets, but it's definitely in the 1000s of messages area when it processes it.  So there is a big spike in activity that is produced when 1 job comes through.  But I'm not sure that job is even making it to that stage!  I found an exception in main which kicks it off when it went to post the first big job.
